tianshou/env/vecenv/shmemenv.py

```diff
@@ -0,0 +1,175 @@
+        self.closed = False
+        dummy = env_fns[0]()
+        obs_space = dummy.observation_space
+        dummy.close()
+        del dummy
+        # will copy be quicker?
+        self.obs_bufs = [self._setup_buf(obs_space)
+                         for _ in range(self.env_num)]
+        self.parent_remote, self.child_remote = \
+            zip(*[Pipe() for _ in range(self.env_num)])
+        self.processes = [
+            Process(target=_shmem_worker, args=(
+                parent, child, CloudpickleWrapper(env_fn),
+                obs_buf), daemon=True)
+            for (parent, child, env_fn, obs_buf) in zip(
+                self.parent_remote, self.child_remote, env_fns, self.obs_bufs)
+        ]
+        for p in self.processes:
+            p.start()
+        for c in self.child_remote:
+            c.close()
+
+    def step(self,
+             action: np.ndarray,
+             id: Optional[Union[int, List[int]]] = None
+             ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+        if id is None:
+            id = range(self.env_num)
+        elif np.isscalar(id):
+            id = [id]
+        assert len(action) == len(id)
+        for i, j in enumerate(id):
+            self.parent_remote[j].send(['step', action[i]])
+        result = []
+        for i in id:
+            obs, rew, done, info = self.parent_remote[i].recv()
+            obs = self._decode_obs(obs, i)
+            result.append((obs, rew, done, info))
+        obs, rew, done, info = map(np.stack, zip(*result))
+        return obs, rew, done, info
+
+    def reset(self, id: Optional[Union[int, List[int]]] = None) -> np.ndarray:
+        if id is None:
+            id = range(self.env_num)
+        elif np.isscalar(id):
+            id = [id]
+        for i in id:
+            self.parent_remote[i].send(['reset', None])
+        obs = np.stack(
+            [self._decode_obs(self.parent_remote[i].recv(), i) for i in id])
+        return obs
+
+    def _setup_buf(self, space):
+        if isinstance(space, gym.spaces.Dict):
+            assert isinstance(space.spaces, OrderedDict)
+            buffer = {k: self._setup_buf(v) for k, v in space.spaces.items()}
+        elif isinstance(space, gym.spaces.Tuple):
+            assert isinstance(space.spaces, tuple)
+            buffer = tuple([self._setup_buf(t) for t in space.spaces])
+        else:
+            buffer = ShArray(space.dtype, space.shape)
+        return buffer
+
+    def _decode_obs(self, isNone, index):
+        def decode_obs(buffer):
+            if isinstance(buffer, ShArray):
+                return buffer.get()
+            elif isinstance(buffer, tuple):
+                return tuple([decode_obs(b) for b in buffer])
+            elif isinstance(buffer, dict):
+                return {k: decode_obs(v) for k, v in buffer.items()}
+            else:
+                raise NotImplementedError
+        assert isNone is None
```

Comment:
I'm puzzled. What does this mean? Why use an argument that is always None?

Reply:
Like 'return None' in encoder, to help people understand that this is a 'fake encoder decoder system', which only buffer the data but not really handle them.

Reply:
> Like 'return None' in encoder, to help people understand that this is a 'fake encoder decoder system', which only buffer the data but not really handle them.

That's crazy, don't do that. This is an object, ``self.decode(i)`` means to decode something inside ``self``, people all know that.

Reply:
Okay, I know it. This parameter is filled by ``recv()`` which returns None. Okay, then.

Reply:
I don't think this is inappropriate, typically decoder receives an object, decodes it return this object. In the current scenario, decoders don't actually have to receive anything. (If it only receives i, this should really be called something like get_buffer_from, not 'decoder').
assert None is a little bit redundant， I can remove that.